[AOTInductor] Support quantized linear on CPU with fbgemm

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122939

Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:
@diff-train-skip-merge
cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

muchulee8 committed:
[WIP] [AOTInductor] Support quantized linear on CPU with fbgemm

Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

muchulee8 committed:
Update on "[WIP] [AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

muchulee8 committed:
Update on "[WIP] [AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

muchulee8 committed:
Update on "[AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

muchulee8 committed:
Update on "[AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -665,6 +667,35 @@ AOTITorchError aoti_torch_mm_out(
+  auto packed_weight = at::fbgemm_pack_gemm_matrix_fp16(*weight_tensor);
+
+  at::Tensor* out_tensor = new at::Tensor(std::move(packed_weight));
+  *out = tensor_pointer_to_tensor_handle(out_tensor);
```
Can we use ```new_tensor_handle```?

muchulee8 replied:
Done.s

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -665,6 +667,35 @@ AOTITorchError aoti_torch_mm_out(
+      *input_tensor, *weight_tensor, *bias_tensor);
+
+  at::Tensor* out_tensor = new at::Tensor(std::move(out_result));
+  *out = tensor_pointer_to_tensor_handle(out_tensor);
```
Use new_tensor_handle?

muchulee8 replied:
Done.

muchulee8 committed:
Update on "[AOTInductor] Support quantized linear on CPU with fbgemm"


Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

chenyang78 reviewed (approved):
LGTM. Thanks!

clee2000 commented:
@pytorchbot merge -f "merged internally"
